packages/nimble-components/build/generate-workers/source/matrix-renderer.ts

```diff
@@ -8,16 +9,158 @@ import { expose } from 'comlink';
+    }
+
+    public setCanvasCorners(topLeft: { x: number, y: number }, bottomRight: { x: number, y: number }): void {
+        this.topLeftCanvasCorner = topLeft;
+        this.bottomRightCanvasCorner = bottomRight;
+    }
+
+    public setDiesDimensions(data: Dimensions): void {
+        this.dieDimensions = { width: data.width, height: data.height };
+    }
+
+    public setScaling(scaleX: number, scaleY: number): void {
+        this.scaleX = scaleX;
+        this.scaleY = scaleY;
+    }
+
+    public setBases(baseX: number, baseY: number): void {
+        this.baseX = baseX;
+        this.baseY = baseY;
+    }
+
+    public setTransform(transform: Transform): void {
+        this.transform = transform;
+    }
+
+    public setCanvas(canvas: OffscreenCanvas): void {
+        this.canvas = canvas;
+        this.context = canvas.getContext('2d')!;
+    }
+
+    public getMatrix(): WaferMapTypedMatrix {
+        return {
+            colIndexes: this.colIndexes,
+            rowIndexes: this.rowIndexes,
+            values: this.values
+        };
+    }
 
     public emptyMatrix(): void {
-        this.dieMatrix = Uint8Array.from([]);;
+        this.colIndexes = Int32Array.from([]);
+        this.rowIndexes = Int32Array.from([]);
+        this.values = Float64Array.from([]);
+    }
+
+    public scaleCanvas(): void {
+        this.context.translate(
+            this.transform.x,
+            this.transform.y
+        );
+        this.context.scale(
+            this.transform.k,
+            this.transform.k
+        );
     }
 
     public updateMatrix(
-        data: Iterable<number>
+        data: WaferMapMatrix
     ): void {
-        this.dieMatrix = Uint8Array.from(data);
+        this.colIndexes = Int32Array.from(data.colIndexes);
+        this.rowIndexes = Int32Array.from(data.rowIndexes);
+        this.values = Float64Array.from(data.values);
+    }
+
+    public setCanvasDimensions(data: Dimensions): void {
+        this.canvas.width = data.width;
+        this.canvas.height = data.height;
+    }
+
+    public clearCanvas(): void {
+        this.context.clearRect(
+            0,
+            0,
+            this.canvas.width,
+            this.canvas.height
+        );
+    }
+
+    public drawWafer(): void {
+        this.context.restore();
+        this.context.save();
+        this.clearCanvas();
+        this.scaleCanvas();
+        if (this.topLeftCanvasCorner === undefined || this.bottomRightCanvasCorner === undefined) { return; }
+        for (let i = 0; i < this.scaledColIndex.length; i++) {
+            const scaledX = this.scaledColIndex[i]!;
+            if (
+                !(scaledX >= this.topLeftCanvasCorner.x
+                    && scaledX < this.bottomRightCanvasCorner.x)
+            ) {
+                continue;
+            }
+            for (let j = this.colIndexPositions[i]!,
+                length = this.colIndexPositions[i + 1] !== undefined ? this.colIndexPositions[i + 1]! : this.scaledRowIndex.length;
```

Comment:
<nit> I think I would prefer that we pull the `length` variable out of the `for` statement, and maybe rename it so that it is less generic. I'll mention that I don't think this is very easy to understand. As written, it appears that the values in `this.colIndexPositions` have two meanings. One is as an index into `this.scaledRowIndex` (line 152), and the other as a _possible_ length to iterate over. Can you explain this? Perhaps in a code comment?

Reply:
I named j and length into columnStartIndex and columnEndIndex and pulled columnEndIndex outside the for, will add a comment tommorow with more details